### PR TITLE
Draft: Add compat atomic header from dav1d

### LINKED ISSUE #1920: Basic atomic operation support
There are a lot of variables declared as volatile, but they don't be access via atomic_load/atomic_store. Looks like there is no basic atomic operation support. How to add the support when C++11 is unavailable?
1. Use compiler extensions?
2. Implement SrtAtomic by Mutex?

### PR DESCRIPTION
**Pull request by quink-black**
Fix #1920

### CODE REVIEW

File: srtcore/atomic_msvc.h [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+#ifndef MSCVER_STDATOMIC_H_
+#define MSCVER_STDATOMIC_H_
+
+#if defined(_MSC_VER)
+
+#pragma warning(push)
+#pragma warning(disable:4067)    /* newline for __has_include_next */
+
+#if defined(__clang__) && __has_include_next(<stdatomic.h>)
```
Comment (maxsharabayko) [inline srtcore/atomic_msvc.h:35]: `clang` in `msvc`... 🤔 
I wonder what is the use case when both are defined?
Maybe clang targeting Window 10 ([CLang MSVC compatibility](https://clang.llvm.org/docs/MSVCCompatibility.html)). But not sure if `_MSC_VER` will be defined.

File: srtcore/atomic_msvc.h [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+*
+* Redistribution and use in source and binary forms, with or without
+* modification, are permitted provided that the following conditions are met:
+*
+* 1. Redistributions of source code must retain the above copyright notice, this
+*    list of conditions and the following disclaimer.
+*
+* 2. Redistributions in binary form must reproduce the above copyright notice,
+*    this list of conditions and the following disclaimer in the documentation
+*    and/or other materials provided with the distribution.
```
Comment (maxsharabayko) [inline srtcore/atomic_msvc.h:13]: Worth checking what do we have to do.
Would we need to put "Copyright © 2018, VideoLAN" somewhere else, e.g. in our [LICENCE](https://github.com/Haivision/srt/blob/master/LICENSE).

**Review by maxsharabayko — COMMENTED** [review]
Wow, @quink-black,  that is super helpful!
This PR provides three possible sources of atomic:
-  `C++11 `std::atomic` when C++11 is enabled;
- [GCC atomic built-ins](https://gcc.gnu.org/onlinedocs/gcc/_005f_005fatomic-Builtins.html). Clang probably goes here as well.
- MSVC atomic via [`InterlockedExchange`](https://docs.microsoft.com/en-us/windows/win32/api/winnt/nf-winnt-interlockedexchange)

There are other possible compilers like IBM, Intel, etc. We don't have to cover every, but it might be worth having some fallback implementation with mutex lock and unlock. And a compilation warning to better enable C++11 for those compilers.
Or just demand to use C++11 when compilers other than gcc, MSVC or clang are used.
Something to decide on.

**Comment by ethouris** [comment]
There is already a PR that includes atomic support and it's one of those submitted for fixing thread sanitizer reported problems.

**Comment by maxsharabayko** [comment]
@ethouris It would be helpful if you could share a link to the PR you mention.

**Comment by ethouris** [comment]
#1863 is the exact PR for atomic, but it would be good if you review everything that is submitted for Epic #1813.

**Comment by quink-black** [comment]
> #1863 is the exact PR for atomic, but it would be good if you review everything that is submitted for Epic #1813.

Good to know that.

dav1d use BSD license, simple google search shows that it's compatible with MPL, but I'm not sure. I have found the same project as in #1863 says:
https://github.com/mbitsnbites/atomic
It doesn't get update since 2017. dav1d is in active develop but it may have different compiler requirements than libsrt. My patch is incomplete since it doesn't take counting of the namespace issue (std::atomic_int vs ::atomic_int).

Feel free to close the PR. I'd like to get more ideas.

**Pull request closed without merging**